pkg/ingester/circuitbreaker_test.go

```diff
@@ -0,0 +1,686 @@
+				# HELP cortex_ingester_circuit_breaker_transitions_total Number of times the circuit breaker has entered a state.
+				# TYPE cortex_ingester_circuit_breaker_transitions_total counter
+				cortex_ingester_circuit_breaker_transitions_total{ingester="ingester",state="closed"} 0
+				cortex_ingester_circuit_breaker_transitions_total{ingester="ingester",state="half-open"} 0
+				cortex_ingester_circuit_breaker_transitions_total{ingester="ingester",state="open"} 1
+				# HELP cortex_ingester_circuit_breaker_current_state Boolean set to 1 whenever the circuit breaker is in a state corresponding to the label name.
+				# TYPE cortex_ingester_circuit_breaker_current_state gauge
+				cortex_ingester_circuit_breaker_current_state{state="open"} 1
+				cortex_ingester_circuit_breaker_current_state{state="half-open"} 0
+				cortex_ingester_circuit_breaker_current_state{state="closed"} 0
+			`,
+		},
+		"if circuit breaker half-open, a circuitBreakerErrorOpen is returned": {
+			circuitBreakerSetup:         func(cb *circuitBreaker) { cb.HalfOpen() },
+			expectedCircuitBreakerError: true,
+			expectedMetrics: `
+				# HELP cortex_ingester_circuit_breaker_results_total Results of executing requests via the circuit breaker.
+				# TYPE cortex_ingester_circuit_breaker_results_total counter
+				cortex_ingester_circuit_breaker_results_total{ingester="ingester",result="circuit_breaker_open"} 1
+				# HELP cortex_ingester_circuit_breaker_transitions_total Number of times the circuit breaker has entered a state.
+				# TYPE cortex_ingester_circuit_breaker_transitions_total counter
+				cortex_ingester_circuit_breaker_transitions_total{ingester="ingester",state="closed"} 0
+				cortex_ingester_circuit_breaker_transitions_total{ingester="ingester",state="half-open"} 1
+				cortex_ingester_circuit_breaker_transitions_total{ingester="ingester",state="open"} 0
+				# HELP cortex_ingester_circuit_breaker_current_state Boolean set to 1 whenever the circuit breaker is in a state corresponding to the label name.
+				# TYPE cortex_ingester_circuit_breaker_current_state gauge
+				cortex_ingester_circuit_breaker_current_state{state="open"} 0
+				cortex_ingester_circuit_breaker_current_state{state="half-open"} 1
+				cortex_ingester_circuit_breaker_current_state{state="closed"} 0
+			`,
+		},
+	}
+	ctx := context.Background()
+	for testName, testCase := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			registry := prometheus.NewRegistry()
+			cb := newCircuitBreaker(cfg, "ingester", log.NewNopLogger(), registry)
+			testCase.circuitBreakerSetup(cb)
+			err := cb.tryAcquirePermit()
+			if testCase.expectedCircuitBreakerError {
+				checkCircuitBreakerOpenErr(ctx, err, t)
+				assert.NoError(t, testutil.GatherAndCompare(registry, strings.NewReader(testCase.expectedMetrics), metricNames...))
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestCircuitBreaker_RecordSuccess(t *testing.T) {
```

Comment:
This test and next one can probably be merged and test both cases at once.

Reply:
Both tests has been removed, and another test testing the newly added method `recordResult()` has been added.